### ISSUE TITLE
PHP 7.2: test suite errors

### ISSUE THREAD

**Issue by zerocrates** [ezyang/htmlpurifier#158]
simpletest sometimes uses old-style "same name as class" constructors, which PHP 7.2 now warns about. This causes one test failure:

> Unexpected PHP Error [Methods with the same name as their class will not be constructors in a future version of PHP; HTMLPurifier_DefinitionTestable has a deprecated constructor] severity [8192] in [simpletest/simpletest/mock_objects.php(1429) : eval()'d code line 1]
> 	in test_get
> 	in HTMLPurifier_DefinitionCache_Decorator_CleanupTest

(Additionally, if running with error_reporting at E_ALL, 4 Deprecated: warnings are printed at test suite startup with the same cause, but those don't cause test failures.)

The underlying issue seems to be in simpletest, so I reported that there as simpletest/simpletest#55

There's also one more "autoload" type problem that appears with E_ALL as well, because tests/common.php has its own separate autoloader. Since it's during test suite startup only, it doesn't cause a failure, though:

> Deprecated: __autoload() is deprecated, use spl_autoload_register() instead in HtmlPurifier/tests/common.php on line 10

**Comment by zerocrates** [comment]
This was caused by me using the simpletest from Composer (side note, if the latest Git is actually required, as some of the code seems to state, then maybe the Composer dev dependency should be on `dev-master`?)

However, using that version causes new different problems, owing to different issues with mock objects, this time syntax errors.

**Comment by zerocrates** [comment]
The new, more accurate for less ancient SimpleTest bug report for the syntax error is simpletest/simpletest#56

There's _another_ problem when that's fixed: partial "testable" mock objects seem to keep the `abstract` in the signature if an abstract method is mocked, leading to this error:

> Fatal error:  Abstract function HTMLPurifier_AttrDefTestable::validate cannot contain body in .../simpletest/mock_objects.php(1394) : eval()'d code on line 138

All that stuff seems to come from one group of reflection changes from December 22, 2016, so maybe the sweet spot is a git checkout _newer_ than the ancient latest release but _older_ than master.

I now notice that you have a forked simpletest which is what's used for Travis and is at exactly the kind of timeframe I was thinking of, so that's probably the somewhat convoluted answer. This doesn't make running the tests particularly easy, I will say.

**Comment by ezyang** [comment]
Yeah, sounds like we should fork the last OK simpletest release and update our code to use it.

**Comment by bytestream** [comment]
@ezyang when do you expect a new release with php 7.2 support? It looks like PR's have already been merged for the issues I've noticed.
